README.md
# The-Bulletin-Board
An incremental game with post-it notes.


changelog.md
# CHANGE LOG

## 19 May 2021

- Game exists
- Implemented basic gameplay
- Added a saving system
- Looking cool

### 21 May 2021

- Color chances can no longer go below 1%

## 21 May 2021 #2

- Numbers below 1 now have 2 significant digits
- Made gray notes slightly lighter
- Slightly changed the appearance of colored text
- Added a bunch more upgrades (too lazy to count them)

#### 22 May 2021

- Upgrade descriptions now have colored text

## 3 Jun. 2021

- Added statistics
- Added achievements, 40 of them
  - No actual use, just there to look cool
- Added a tab icon

### 9 Sep. 2021

- Changed around some of the prices

### 9 Sep. 2021 #2

- Fixed a bug where post interval upgrades won't appear

### 9 Sep. 2021 #3

- Added a new green upgrade

### 9 Sep. 2021 #4

- Fixed a bug where offline progress will also spawn posts


js/saving.js
function save(auto = true) {
  player.date = Date.now();
  let data = JSON.stringify(player);
  localStorage.setItem("tbbsave", data);
  if (!auto) dropPost("saveBtn", false);
}
function load(auto = true) {
  let data = JSON.parse(localStorage.getItem("tbbsave"));
  if (data) {
    let dt = player.date - data.date;
    noPost = true;
    simulateTime(dt);
    noPost = false;
    data.date = player.date;
    player.board = [...data.board];
    if (data.achievements) player.achievements = [...data.achievements];
    merge(player, data);
    if (!auto) dropPost("loadBtn", false);
  }
  updateDisp();
  updateBoard();
  updateChance();
  updateBarLabel();
  updateSettings();
  updateStats();
  updateAch();
}
function exportSave() {
  let data = JSON.stringify(player);
  data = btoa(data);
  id("exportArea").classList.remove("hidden");
  id("exportArea").innerHTML = data;
  id("exportArea").select();
  document.execCommand("copy");
  id("exportArea").classList.add("hidden");
  dropPost("exportBtn", false);
}
function importSave() {
  let data = prompt("Please enter exported save:");
  if (data) {
    try {
      data = atob(data);
      JSON.parse(data);
      localStorage.setItem("tbbsave", data);
      load();
      dropPost("importBtn", false);
    } catch (err) {
      alert("Invalid save");
    }
  }
}
function toggleAutoSave() {
  player.autoSave = !player.autoSave;
  dropPost("autoSaveBtn", false);
  setTimeout(() => {
    updateSettings();
  }, 500);
}
function editSaveInterval() {
  let newInterval = prompt(
    "Please enter the new auto save interval in miliseconds (500 to 86400000 inclusive):"
  );
  if (newInterval) {
    if (newInterval >= 500 && newInterval <= 86400000) {
      player.autoSaveInterval = newInterval;
      setTimeout(() => {
        updateSettings();
      }, 500);
      dropPost("saveIntervalBtn", false);
    } else alert("Out of range!");
  }
}
function wipeSave() {
  player = newGame();
  save();
  load();
  nextPost = 0;
  dropPost("wipeSaveBtn", false);
}
function wipeSaveConfirm() {
  if (confirm("Are you sure you want to wipe all save data?")) wipeSave();
}
function merge(base, source) {
  for (let i in base) {
    if (source[i] !== undefined) {
      if (typeof base[i] === "object" && typeof source[i] === "object") {
        merge(base[i], source[i]);
      } else {
        base[i] = source[i];
      }
    }
  }
}


js/update.js
function updateDisp() {
  let types = ["y", "b", "g", "p"];
  for (let i in types) {
    i = parseInt(i);
    id(types[i] + "Amt").innerHTML = format(player.notes[i]);
    id(types[i] + "Rate").innerHTML = format(getRate(i)) + "/s";
    if (player.notesTotal[i] > 0)
      id(types[i] + "Disp").classList.remove("hidden");
    if (player.doubleTime > 0) {
      id(types[i] + "Rate").classList.add("greenText");
      id(types[i] + "Rate").classList.remove("subtext");
    } else {
      id(types[i] + "Rate").classList.remove("greenText");
      id(types[i] + "Rate").classList.add("subtext");
    }

    if (player.notes[i] >= 1 && [1, 2].includes(i)) giveAch(types[i] + "Note1");
    if (player.notes[i] >= 10 && [0, 1, 2].includes(i))
      giveAch(types[i] + "Note10");
    if (player.notes[i] >= 100 && [0, 1, 2].includes(i))
      giveAch(types[i] + "Note100");
    if (player.notes[i] >= 1000 && [0, 1, 2].includes(i))
      giveAch(types[i] + "Note1000");
    if (player.notes[i] >= 1e4 && [0, 1].includes(i))
      giveAch(types[i] + "Note1e4");
    if (player.notes[i] >= 1e5 && [0].includes(i))
      giveAch(types[i] + "Note1e5");

    if (player.rate[i] >= 0.01 && [1].includes(i))
      giveAch(types[i] + "Rate0.01");
    if (player.rate[i] >= 0.1 && [0, 1].includes(i))
      giveAch(types[i] + "Rate0.1");
    if (player.rate[i] >= 1 && [0, 1].includes(i)) giveAch(types[i] + "Rate1");
    if (player.rate[i] >= 10 && [0, 1].includes(i))
      giveAch(types[i] + "Rate10");
    if (player.rate[i] >= 100 && [0, 1].includes(i))
      giveAch(types[i] + "Rate100");
    if (player.rate[i] >= 1000 && [0].includes(i))
      giveAch(types[i] + "Rate1000");

    if (player.notes[i] < 0) giveAch("negative");
  }
  if (player.doubleTime > 0) {
    id("doubleDisp").classList.remove("hidden");
    id("doubleTimer").innerHTML = formatTime(player.doubleTime);
  } else id("doubleDisp").classList.add("hidden");
}

function updateChance() {
  let text = "Chances:";
  if (player.noteChance[0] !== 0)
    text += `<br><span class='yellowText'>${format(
      player.noteChance[0] * 100
    )}%</span>`;
  if (player.noteChance[1] !== 0)
    text += `<br><span class='blueText'>${format(
      player.noteChance[1] * 100
    )}%</span>`;
  if (player.noteChance[2] !== 0)
    text += `<br><span class='greenText'>${format(
      player.noteChance[2] * 100
    )}%</span>`;
  if (player.noteChance[3] !== 0)
    text += `<br><span class='purpleText'>${format(
      player.noteChance[3] * 100
    )}%</span>`;
  id("chanceDisp").innerHTML = text;
}

function updateBoard() {
  let types = ["yellow", "blue", "green", "purple"];
  for (let i in player.board) {
    let postId = player.board[i];
    if (postId === "empty") {
      id("slot" + i).classList.add("hidden");
      continue;
    }
    let post = noteData[postId[0]][postId[1]];
    let text = post.desc
      .replaceAll("<y>", "<span class='yellowText'>")
      .replaceAll("<b>", "<span class='blueText'>")
      .replaceAll("<g>", "<span class='greenText'>")
      .replaceAll("<p>", "<span class='purpleText'>")
      .replaceAll("</>", "</span>");
    if (post.cost.some((x) => x !== 0)) {
      text += "<br>Cost: ";
      if (post.cost[0] !== 0)
        text += `<span class='yellowText'>${format(post.cost[0])}</span> `;
      if (post.cost[1] !== 0)
        text += `<span class='blueText'>${format(post.cost[1])}</span> `;
      if (post.cost[2] !== 0)
        text += `<span class='greenText'>${format(post.cost[2])}</span> `;
      if (post.cost[3] !== 0)
        text += `<span class='purpleText'>${format(post.cost[3])}</span>`;
    }
    id("slot" + i).classList.remove(
      "hidden",
      "used",
      "blue",
      "green",
      "purple"
    );
    id("slot" + i).classList.add(types[post.type]);
    id("text" + i).innerHTML = text;
  }
}

function updateBar() {
  id("barFill").style.height = `${(nextPost / player.postInterval) * 100}%`;
}

function updateBarLabel() {
  id("barDisp").innerHTML = `${getFilled()}/${player.boardSpace}`;
  if (getFilled() === player.boardSpace) {
    id("barDisp").style.color = "red";
  } else id("barDisp").style.color = "";
}

function updateSettings() {
  id("autoSaveDisp").innerHTML = player.autoSave ? "ON" : "OFF";
  id("saveIntervalDisp").innerHTML = formatTime(player.autoSaveInterval);
}

function updateStats() {
  id("timePlayed").innerHTML = formatTime(player.timePlayed);
  id("postsUsed").innerHTML = player.postsUsed;
  id("postsDiscarded").innerHTML = player.postsDiscarded;
  let types = ["Y", "B", "G", "P"];
  for (let i in types) {
    id("totalNotes" + types[i]).innerHTML = format(player.notesTotal[i]);
    if (player.notesTotal[i] > 0)
      id("totalNotes" + types[i] + "Container").classList.remove("hidden");
  }
}

function updateAch() {
  for (let i in player.achievements) {
    let achId = player.achievements[i];
    let ach = achList[achId];
    id(achId + "Ach").classList.remove("gray");
    id(
      achId + "Ach"
    ).innerHTML = `<span>${ach.name}<br><span class="subtext">${ach.desc}</span></span>`;
  }
}

var notifyQueue = [];
function notify(title, text) {
  notifyQueue.push([title, text]);
  updateNotification();
}

function updateNotification() {
  let data = notifyQueue[0];
  if (data) {
    let text = `<span>${data[0]}<br><span class="subtext">${data[1]}</span></span>`;
    id("notifications").innerHTML = text ?? "";
  }
  id("notifications").classList[notifyQueue.length === 0 ? "add" : "remove"](
    "hidden"
  );
}

function removeNotification() {
  notifyQueue.shift();
  updateNotification();
}


js/upgData.js
const noteData = [[], [], [], []];

class Upg {
  constructor(type, cost, weight, desc, effect, arg, condition) {
    noteData[type].push(this);
    this.type = type;
    this.cost = cost;
    this.weight = weight;
    this.desc = desc;
    this.effect = effect;
    this.arg = arg;
    this.condition = condition;
  }
}

// yellow
new Upg(
  0,
  [0, 0, 0, 0],
  10,
  "+1 <y>note</>",
  getNotes,
  [0, 1],
  () => player.rate[0] < 2
);
new Upg(
  0,
  [0, 0, 0, 0],
  10,
  "+2 <y>notes</>",
  getNotes,
  [0, 2],
  () => player.notes[0] >= 5 && player.rate[0] < 4
);
new Upg(
  0,
  [0, 0, 0, 0],
  5,
  "+5 <y>notes</>",
  getNotes,
  [0, 5],
  () => player.notes[0] >= 20 && player.rate[0] < 10
);
new Upg(
  0,
  [50, 0, 0, 0],
  2,
  "+1 <b>blue note</>",
  getNotes,
  [1, 1],
  () => player.notes[0] >= 25 && player.rate[1] < 2
);
new Upg(
  0,
  [150, 0, 0, 0],
  1,
  "+2 <b>blue notes</>",
  getNotes,
  [1, 2],
  () => player.notes[1] >= 5 && player.rate[1] < 4
);
new Upg(
  0,
  [350, 0, 0, 0],
  0.5,
  "+5 <b>blue notes</>",
  getNotes,
  [1, 5],
  () => player.notes[1] >= 20 && player.rate[1] < 10
);
new Upg(
  0,
  [1000, 50, 0, 0],
  2,
  "+1 <g>green note</>",
  getNotes,
  [2, 1],
  () => player.notes[1] >= 25 && player.rate[2] < 2
);
new Upg(
  0,
  [2500, 150, 0, 0],
  1,
  "+2 <g>green notes</>",
  getNotes,
  [2, 2],
  () => player.notes[2] >= 5 && player.rate[2] < 4
);
new Upg(
  0,
  [6500, 350, 0, 0],
  0.5,
  "+5 <g>green notes</>",
  getNotes,
  [2, 5],
  () => player.notes[2] >= 20 && player.rate[2] < 10
);
new Upg(
  0,
  [0, 0, 0, 0],
  50,
  "Reset <y>notes</> & <y>note/s</> at the cost of all <b>blue note/s</>",
  () => {
    player.notes[0] = 0;
    player.rate[0] = 0;
    player.rate[1] = 0;
  },
  [],
  () => player.rate[0] < 0 && player.notes[0] < -10
);
new Upg(
  0,
  [0, 0, 0, 0],
  0.05,
  "Gain <y>notes</> equal to 30 minute of production",
  getNotes,
  [0, () => player.rate[0] * 60 * 30],
  () => player.rate[0] > 0
);
new Upg(
  0,
  [0, 0, 0, 0],
  0.05,
  "Gain <b>blue notes</> equal to 30 minute of production",
  getNotes,
  [1, () => player.rate[1] * 60 * 30],
  () => player.rate[1] > 0
);
new Upg(
  0,
  [0, 0, 0, 0],
  0.1,
  "Does nothing",
  () => {},
  [],
  () => true
);

// blue
new Upg(
  1,
  [5, 0, 0, 0],
  10,
  "+0.1 <y>notes/s</>",
  addRate,
  [0, 0.1],
  () => player.rate[0] < 1
);
new Upg(
  1,
  [15, 0, 0, 0],
  5,
  "+0.2 <y>notes/s</>",
  addRate,
  [0, 0.2],
  () => player.rate[0] >= 0.5 && player.rate[0] < 2
);
new Upg(
  1,
  [50, 1, 0, 0],
  4,
  "+0.5 <y>notes/s</>",
  addRate,
  [0, 0.5],
  () => player.rate[0] >= 2
);
new Upg(
  1,
  [20, 0, 0, 0],
  3,
  "+5% <b>blue chance</>, -5% <y>yellow chance</>",
  chanceExch,
  [0, 1, 0.05],
  () => player.notes[0] >= 20 && player.noteChance[0] > 0.01
);
new Upg(
  1,
  [20, 0, 0, 0],
  3,
  "+5% <y>yellow chance</>, -5% <b>blue chance</>",
  chanceExch,
  [1, 0, 0.05],
  () => player.notes[0] >= 20 && player.noteChance[1] > 0.01
);
new Upg(
  1,
  [50, 5, 0, 0],
  1,
  "+5% <g>green chance</>, -5% <b>blue chance</>",
  chanceExch,
  [1, 2, 0.05],
  () => player.notes[1] >= 5 && player.noteChance[1] > 0.01
);
new Upg(
  1,
  [50, 5, 0, 0],
  1,
  "+5% <b>blue chance</>, -5% <g>green chance</>",
  chanceExch,
  [2, 1, 0.05],
  () => player.notes[1] >= 5 && player.noteChance[2] > 0.01
);
new Upg(
  1,
  [200, 5, 0, 0],
  2,
  "+1 <y>notes/s</>",
  addRate,
  [0, 1],
  () => player.rate[0] >= 5
);
new Upg(
  1,
  [1000, 25, 1, 0],
  2,
  "+5 <y>notes/s</>",
  addRate,
  [0, 5],
  () => player.rate[0] >= 25
);
new Upg(
  1,
  [0, 0, 0, 0],
  0.1,
  "Does nothing",
  () => {},
  [],
  () => true
);
// green
new Upg(
  2,
  [0, 1, 0, 0],
  10,
  "Gain <y>notes</> equal to current <b>blue notes</> * 10",
  getNotes,
  [0, () => player.notes[1] * 10],
  () => true
);
new Upg(
  2,
  [0, 0, 1, 0],
  2,
  "Convert ALL <y>notes</> to <b>blue notes</> at a <y>200</>:<b>1</> ratio.",
  () => {
    getNotes(1, player.notes[0] / 200);
    getNotes(0, -player.notes[0]);
  },
  [],
  () => player.notes[0] >= 1e4
);
new Upg(
  2,
  [250, 5, 0, 0],
  5,
  "+0.01 <b>blue note/s</>",
  addRate,
  [1, 0.01],
  () => player.notes[1] >= 10 && player.rate[1] < 0.1
);
new Upg(
  2,
  [500, 15, 0, 0],
  5,
  "+0.02 <b>blue note/s</>",
  addRate,
  [1, 0.02],
  () => player.rate[1] >= 0.05 && player.rate[1] < 0.2
);
new Upg(
  2,
  [1500, 35, 1, 0],
  5,
  "+0.05 <b>blue note/s</>",
  addRate,
  [1, 0.05],
  () => player.rate[1] >= 0.2 && player.rate[1] < 0.5
);
new Upg(
  2,
  [3000, 100, 1, 0],
  2,
  "Decrease post spawn interval by 100 ms",
  addPostDelay,
  [-100],
  () => player.postInterval > 500
);
new Upg(
  2,
  [3000, 100, 1, 0],
  2,
  "Increase post spawn interval by 100 ms",
  addPostDelay,
  [100],
  () => player.postInterval < 1000 * 60
);
new Upg(
  2,
  [1e4, 250, 3, 0],
  1,
  "+1 board space (up to 10)",
  expandBoard,
  [1, 10],
  () => player.boardSpace < 10
);
new Upg(
  2,
  [0, 0, 1, 0],
  3,
  "Double all production for 1 minute",
  setDoubleTime,
  [60000],
  () => player.notes[2] >= 1 && player.doubleTime <= 0
);
new Upg(
  2,
  [2000, 90, 1, 0],
  2.5,
  "+0.1 <b>blue note/s</> but -2 <y>note/s</>",
  () => {
    addRate(1, 0.1);
    addRate(0, -2);
  },
  [],
  () => player.rate[1] >= 0.1 && player.rate[0] >= 2 && player.rate[1] < 1
);
new Upg(
  2,
  [4000, 200, 2, 0],
  2.5,
  "+0.2 <b>blue note/s</> but -5 <y>note/s</>",
  () => {
    addRate(1, 0.2);
    addRate(0, -5);
  },
  [],
  () => player.rate[1] >= 0.2 && player.rate[0] >= 5 && player.rate[1] < 2
);
new Upg(
  2,
  [10000, 500, 4, 0],
  2.5,
  "+0.5 <b>blue note/s</> but -15 <y>note/s</>",
  () => {
    addRate(1, 0.5);
    addRate(0, -15);
  },
  [],
  () => player.rate[1] >= 0.5 && player.rate[0] >= 15
);
new Upg(
  2,
  [0, 0, 0, 0],
  0.1,
  "Does nothing",
  () => {},
  [],
  () => true
);
// purple

function getTotalWeight(type) {
  let sum = 0;
  for (let i in noteData[type]) {
    if (noteData[type][i].condition()) sum += noteData[type][i].weight;
  }
  return sum;
}


js/achievements.js
var achList = [];

class Achievement {
  constructor(achId, name, desc) {
    achList[achId] = this;
    this.name = name;
    this.desc = desc;
    let disp = document.createElement("div");
    disp.id = achId + "Ach";
    disp.innerHTML = `<span>???<br><span class="subtext">???</span></span>`;
    disp.classList.add("post", "gray");
    disp.style.position = "relative";
    id("achievementDisp").appendChild(disp);
  }
}

function giveAch(id) {
  if (!player.achievements.includes(id)) {
    player.achievements.push(id);
    updateAch();
    notify("Achievement Got!", achList[id].name);
  }
}

new Achievement("start", "Great Start!", "Start the game.");
// yellow note collection
new Achievement("yNote10", "Sticky Fingers", "Gain a total of 10 notes.");
new Achievement("yNote100", "Enough for a Pack", "Gain a total of 100 notes.");
new Achievement("yNote1000", "Tiny Factory", "Gain a total of 1000 notes.");
new Achievement(
  "yNote1e4",
  "It's Not Raining Water",
  "Gain a total of 1e4 notes."
);
new Achievement("yNote1e5", "Pastel El Dorado", "Gain a total of 1e5 notes.");
// blue note collection
new Achievement("bNote1", "Ocean Blue", "Gain a total of 1 blue note.");
new Achievement(
  "bNote10",
  "Just a Bit of Variety",
  "Gain a total of 10 blue note."
);
new Achievement(
  "bNote100",
  "Yellow Is Boring",
  "Gain a total of 100 blue notes."
);
new Achievement("bNote1000", "Ya Blue It", "Gain a total of 1000 blue notes.");
new Achievement("bNote1e4", "Pacific Paper", "Gain a total of 1e4 blue notes.");
// green note collection
new Achievement(
  "gNote1",
  "Green Like the Trees",
  "Gain a total of 1 green note."
);
new Achievement("gNote10", "10 Leaf Clover", "Gain a total of 10 green notes.");
new Achievement(
  "gNote100",
  "Wished It Was $100",
  "Gain a total of 100 green notes."
);
new Achievement(
  "gNote1000",
  "Processed Forest",
  "Gain a total of 1000 green notes."
);
// yellow note rate
new Achievement("yRate0.1", "Automation!", "Produce 0.1 notes/s.");
new Achievement("yRate1", "It's Whole Now", "Produce 1 note/s.");
new Achievement("yRate10", "Factory in the Making", "Produce 10 notes/s.");
new Achievement("yRate100", "AI Revolution? Nah", "Produce 100 notes/s.");
new Achievement("yRate1000", "Think of the Trees!", "Produce 1000 notes/s.");
// blue note rate
new Achievement("bRate0.01", "It Flows", "Produce 0.01 blue notes/s.");
new Achievement("bRate0.1", "Catching Up", "Produce 0.1 blue notes/s.");
new Achievement("bRate1", "Drop in the Pond", "Produce 1 blue note/s.");
new Achievement("bRate10", "Steady Hydration", "Produce 10 blue notes/s.");
new Achievement("bRate100", "Dihydrogen Monoxide", "Produce 100 blue notes/s.");
// post used
new Achievement("use10", "Ooh Paper", "Use a total of 10 posts.");
new Achievement("use100", "Post-it, Collect It", "Use a total of 100 posts.");
new Achievement(
  "use1000",
  "Post-It Note Enthusiast",
  "Use a total of 1000 posts."
);
new Achievement(
  "use1e4",
  "Ok Now You're Just Obsessed",
  "Use a total of 1e4 posts."
);
// post discarded
new Achievement("discard10", "Picky", "Discard a total of 10 posts.");
new Achievement("discard100", "Me No Likey", "Discard a total of 100 posts.");
new Achievement(
  "discard1000",
  "Trash Can Free Throw",
  "Discard a total of 1000 posts."
);
new Achievement(
  "discard1e4",
  "I Hope You're Recycling",
  "Discard a total of 1e4 posts."
);
// board space
new Achievement("board6", "Expansion Pack", "Have a board size of 6.");
new Achievement("board10", "I Felt Cramped", "Have a board size of 10.");
// special
new Achievement("openSettings", "Decisions", "Visit the settings.");
new Achievement("openStats", "Numbers n Stuff", "Visit the statistics.");
new Achievement("openAch", "Hello There", "Visit the achievements.");
new Achievement("1%", "Nuh Uh", "Try to go below 1% chance.");
new Achievement(
  "negative",
  "Dipping Into The Unknown",
  "Have negative of any notes."
);


js/base.js
var player = newGame();
const maxBoardSpace = 25;
function newGame() {
  return {
    timePlayed: 0,
    board: [
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty"
    ],
    notes: [0, 0, 0, 0],
    notesTotal: [0, 0, 0, 0],
    postsUsed: 0,
    postsDiscarded: 0,
    rate: [0, 0, 0, 0],
    postInterval: 1000 * 5,
    boardSpace: 5,
    noteChance: [0.8, 0.2, 0, 0],
    date: Date.now(),
    autoSave: true,
    autoSaveInterval: 10000,
    doubleTime: 0,
    achievements: []
  };
}
function getNotes(type, amount) {
  player.notes[type] += amount;
  player.notesTotal[type] += amount;
}

function addRate(type, amount) {
  player.rate[type] += amount;
}

function chanceExch(fromType, toType, amount) {
  if (amount > player.noteChance[fromType] - 0.01) giveAch("1%");
  amount = Math.min(player.noteChance[fromType] - 0.01, amount);
  if (amount < 0) amount = 0;
  player.noteChance[fromType] -= amount;
  player.noteChance[toType] += amount;
  updateChance();
}

function addPostDelay(amount) {
  player.postInterval += amount;
  player.postInterval = Math.max(player.postInterval, 500);
  player.postInterval = Math.min(player.postInterval, 1000 * 60);
}

function expandBoard(amount, limit) {
  if (player.boardSpace < limit) player.boardSpace += amount;
  if (player.boardSpace >= 6) giveAch("board6");
  if (player.boardSpace >= 10) giveAch("board10");
  updateBarLabel();
}

function setDoubleTime(time) {
  player.doubleTime = time;
}

function addPost() {
  let pos = Math.floor(Math.random() * player.boardSpace);
  while (player.board[pos] !== "empty") {
    pos = Math.floor(Math.random() * player.boardSpace);
  }
  let type = Math.random();
  for (let i in player.noteChance) {
    if (type < player.noteChance[i]) {
      type = i;
      break;
    }
    type -= player.noteChance[i];
  }
  let post = Math.random() * getTotalWeight(type);
  for (let i in noteData[type]) {
    if (noteData[type][i].condition()) {
      if (post < noteData[type][i].weight) {
        post = [type, i];
        break;
      }
      post -= noteData[type][i].weight;
    }
  }
  player.board[pos] = post;
  updateBoard();
  updateBarLabel();
}

function usePost(pos) {
  let postId = player.board[pos];
  if (postId !== "empty") {
    let post = noteData[postId[0]][postId[1]];
    let cost = post.cost;
    for (let i in cost) {
      if (player.notes[i] < cost[i] && cost[i] !== 0) break;
      if (i == 3) {
        for (let j in cost) player.notes[j] -= cost[j];
        let arg = deepCopy(post.arg);
        for (let i in arg) {
          if (typeof arg[i] === "function") arg[i] = arg[i]();
        }
        post.effect(...arg);
        player.postsUsed++;
        if (player.postsUsed >= 10) giveAch("use10");
        if (player.postsUsed >= 100) giveAch("use100");
        if (player.postsUsed >= 1000) giveAch("use100");
        if (player.postsUsed >= 1e4) giveAch("use1e4");
        removePost(pos);
      }
    }
  }
}

function removePost(pos) {
  player.board[pos] = "empty";
  dropPost("slot" + pos);
  setTimeout(function () {
    updateBoard();
  }, 500);
  updateBarLabel();
  return false;
}

function dropPost(eid, hide = true) {
  id(eid).classList.add("used");
  if (!hide) {
    setTimeout(() => {
      id(eid).classList.remove("used");
    }, 500);
  }
}

function getFilled() {
  return player.board.filter((x) => x !== "empty").length;
}

function getRate(type) {
  return player.rate[type] * (player.doubleTime > 0 ? 2 : 1);
}

function toggleSettings() {
  if (id("screenContainer").style.transform === "") {
    id("screenContainer").style.transform = "translate(-100vw,0)";
    giveAch("openSettings");
  } else id("screenContainer").style.transform = "";
}
function toggleStats() {
  if (id("screenContainer").style.transform === "") {
    id("screenContainer").style.transform = "translate(100vw,0)";
    giveAch("openStats");
  } else id("screenContainer").style.transform = "";
}
function toggleAchievements() {
  if (id("screenContainer").style.transform === "translate(100vw, 0px)") {
    id("screenContainer").style.transform = "translate(100vw,100vh)";
    giveAch("openAch");
  } else id("screenContainer").style.transform = "translate(100vw,0)";
}

var prevts = 0;
var nextPost = 0;
var nextSave = 0;
var noPost = false;
function nextFrame(ts) {
  if (prevts === 0) prevts = ts;
  let dt = ts - prevts;
  prevts = ts;
  player.timePlayed += dt;
  if (id("screenContainer").style.transform === "translate(100vw, 0px)")
    updateStats();
  simulateTime(dt);
  requestAnimationFrame(nextFrame);
}

function simulateTime(ms) {
  let t = ms / 1000;
  for (let i in player.notes) {
    player.notes[i] += getRate(i) * t;
    player.notesTotal[i] += getRate(i) * t;
  }
  if (!noPost) nextPost += ms;
  if (nextPost > player.postInterval) {
    nextPost %= player.postInterval;
    if (getFilled() === player.boardSpace) {
      removePost(Math.floor(Math.random() * player.boardSpace));
    } else addPost();
  }
  if (player.autoSave) {
    nextSave += ms;
    if (nextSave > player.autoSaveInterval) {
      nextSave %= player.autoSaveInterval;
      save();
    }
  }
  if (player.doubleTime > 0) player.doubleTime -= ms;
  if (player.doubleTime < 0) player.doubleTime = 0;
  updateDisp();
  updateBar();
}

const id = (x) => document.getElementById(x);

for (let i = 0; i < maxBoardSpace; i++) {
  let ele = document.createElement("div");
  ele.id = "slot" + i;
  ele.classList.add("post", "hidden");
  ele.addEventListener("click", function (event) {
    if (event.shiftKey) {
      removePost(i);
      player.postsDiscarded++;
      if (player.postsDiscarded >= 10) giveAch("discard10");
      if (player.postsDiscarded >= 100) giveAch("discard100");
      if (player.postsDiscarded >= 1000) giveAch("discard100");
      if (player.postsDiscarded >= 1e4) giveAch("discard1e4");
    } else usePost(i);
  });
  let text = document.createElement("span");
  text.id = "text" + i;
  ele.appendChild(text);
  id("board").appendChild(ele);
}

document.querySelectorAll(".post").forEach(function (ele) {
  ele.style.transform = `rotate(${Math.random() * 10 - 5}deg)`;
});

function deepCopy(inObject) {
  let outObject, value, key;
  if (typeof inObject !== "object" || inObject === null) {
    return inObject;
  }
  outObject = Array.isArray(inObject) ? [] : {};
  for (key in inObject) {
    value = inObject[key];
    outObject[key] = deepCopy(value);
  }
  return outObject;
}

function format(n) {
  if (n < 1) {
    return n.toFixed(2);
  } else if (n < 1e3) {
    return n.toFixed(1);
  } else {
    return n.toExponential(1).replace("+", "");
  }
}

function formatTime(ms) {
  let s = ms / 1000;
  let ds = s % 60;
  let m = Math.floor(s / 60);
  let dm = m % 60;
  let h = Math.floor(m / 60);
  let dh = h % 24;
  let d = Math.floor(h / 24);
  let dd = d % 30.43685;
  let mo = Math.floor(d / 30.43685);
  let dmo = mo % 12;
  let dy = Math.floor(mo / 365.2422);
  let time =
    s < 60 ? ds.toFixed(2) + "s" : (ds < 10 ? "0" : "") + ds.toFixed(0);
  if (dm >= 1) time = (dm < 10 ? "0" : "") + dm + ":" + time;
  if (dh >= 1) time = (dh < 10 ? "0" : "") + dh + ":" + time;
  if (dd >= 1) time = dd + ":" + time;
  if (dmo >= 1) time = dmo + ":" + time;
  if (dy >= 1) time = dy + ":" + time;
  return time;
}

function init() {
  dropPost("startBtn");
  setTimeout(() => {
    id("startBtnCont").classList.add("hidden");
  }, 500);
  id("topLeft").classList.remove("hidden");
  id("topRight").classList.remove("hidden");
  id("board").classList.remove("hidden");
  id("bottomLeft").classList.remove("hidden");
  id("sideBar").classList.remove("hidden");
  getNotes(0, 1);
  load();
  giveAch("start");
  requestAnimationFrame(nextFrame);
}
